src/hotspot/share/gc/parallel/psAdaptiveSizePolicy.cpp

```diff
@@ -37,6 +37,244 @@
+                              "IOPrioritySizePolicy can't decrease heap below these sizes",
+                              ParallelScavengeHeap::young_gen()->min_gen_size()/M,
+                              ParallelScavengeHeap::old_gen()->min_gen_size()/M);
+  }
+
+  void start_mutator_record() {
+    if (!_active) {
+      return;
+    }
+    _io_wait_record.start();
+    _user_time_record.start();
+  }
+
+  void stop_mutator_record() {
+    if (!_active) {
+      return;
+    }
+    _mutator_iowait_percent = _io_wait_record.stop();
+    _mutator_user_percent = _user_time_record.stop();
+  }
+
+  void print(double mutator_cost) const {
+    if (!_active) {
+      return;
+    }
+    log_debug(gc, ergo, heap)("mutator cost: %f, iowait : %f, user: %f", mutator_cost, _mutator_iowait_percent, _mutator_user_percent);
+  }
+
+  bool should_full_GC() {
+    if (!_active) {
+      return false;
+    }
+
+    // These thresholds are turned by spark on TPC-DS workload.
+    const static double IOTriggerredFullGCUserThreshold = 0.75;
+    const static double IOTriggerredFullGCIOWaitThreshold = 0.4;
+    const static double IOTriggerredFullGCMinInterval = 60; // can be set longer if io heavy workload lasts long.
+
+    if (_mutator_user_percent < IOTriggerredFullGCUserThreshold &&
+        _mutator_iowait_percent > IOTriggerredFullGCIOWaitThreshold) {
+      _io_triggerred_major_gc_timer.stop();
+      if (_io_triggerred_major_gc_timer.seconds() >
+          IOTriggerredFullGCMinInterval) {
+        _io_triggerred_major_gc_timer.reset();
+        _io_triggerred_major_gc_timer.start();
+        log_debug(gc, ergo, heap)("decrease old gen by full gc");
+        return true;
+      } else {
+        log_debug(gc, ergo, heap)(
+            "decrease old gen FAILED because interval is %f < %f",
+            _io_triggerred_major_gc_timer.seconds(), IOTriggerredFullGCMinInterval);
+        _io_triggerred_major_gc_timer.start();
+        return false;
+      }
+    }
+    return false;
+  }
+
+  double calculate_reduced_throughput_goal() {
+    if (!_active) {
+      return _default_throughput_goal;
+    }
+
+    const static double UserThreshold = 1.0;
+    const static double IOWaitThreshold = 0.1;
+
+    if (_mutator_user_percent < UserThreshold &&
+        _mutator_iowait_percent > IOWaitThreshold) {
+      double reduced_throughput_goal = _default_throughput_goal - (1 - _mutator_user_percent);
+      _should_reduce_heap = true;
+      log_debug(gc, ergo, heap)("decrease throughput goal to %.3f", reduced_throughput_goal);
+      return reduced_throughput_goal;
+    } else {
+      _should_reduce_heap = false;
+      return _default_throughput_goal;
+    }
+  }
+
+  size_t calculate_reduced_eden_size(size_t eden_size, float avg_survivor, size_t current_eden_size) const {
+    if (!_active || !_should_reduce_heap) {
+      return eden_size;
+    }
+    size_t reduced_size;
+    reduced_size = MIN(eden_size, avg_survivor * SurvivorRatio);
```

Comment:
On second though, we should probably use another variable, otherwise SurvivorRatio will serve two purposes (original purpose is to control survivor space size; now it's used to control eden size)

This variable could be initially set to the same value as SurvivorRatio

Reply:
When Eden size is reduced, survivor rate might increase due to frequent ygc. So probably consider that and leave some leeway.

Reply:
done